### ISSUE TITLE
InvalidCastException in RemoveUnnecessaryCast analyzer on BoundTupleLiteral

### ISSUE THREAD

**Issue by cincuranet** [dotnet/roslyn#17841]
**Version Used**: 
VS 2017 RTM, Community

**Steps to Reproduce**:

This is the method causing the issue.

```csharp
public static IEnumerator<(int?, string, long?, short?, string, DateTime?, DateTime?, TimeSpan?, bool?, float?, double?, decimal?, decimal?)> Demo(int? i, string s, long? @long, short? @short, string @char, DateTime? timestamp, DateTime? date, TimeSpan? time, bool? boolean, float? @float, double? @double, decimal? @decimal, decimal? numeric)
{
	yield return (i, s, @long, @short, @char, timestamp, date, time, boolean, @float, @double, @decimal, numeric);
	yield return (i + 1, s + nameof(Demo), @long + 1, (short?)(@short + 1), @char.TrimEnd() + "_" + nameof(Demo), timestamp?.AddDays(1).AddHours(1), date?.AddDays(1), time?.Add(TimeSpan.FromMinutes(1)), !boolean, @float + 1.1f, @double + 1.1, @decimal + 1.1m, numeric + 1.1m);
}
```

```
Severity	Code	Description	Project	File	Line	Suppression State	Detail Description
Warning	AD0001	Analyzer 'Microsoft.CodeAnalysis.CSharp.Diagnostics.RemoveUnnecessaryCast.CSharpRemoveUnnecessaryCastDiagnosticAnalyzer' threw an exception of type 'System.InvalidCastException' with message 'Unable to cast object of type 'Microsoft.CodeAnalysis.CSharp.BoundTupleLiteral' to type 'Microsoft.CodeAnalysis.CSharp.BoundConvertedTupleLiteral'.'.	Example		1	Active	Analyzer 'Microsoft.CodeAnalysis.CSharp.Diagnostics.RemoveUnnecessaryCast.CSharpRemoveUnnecessaryCastDiagnosticAnalyzer' threw the following exception:
'Exception occurred with following context:
Compilation: Example
SyntaxTree: C:\Users\Jiri\Documents\devel\JiriPlugin\Example\Procedures.cs
SyntaxNode: (short?)(@short + 1) [CastExpressionSyntax]@[651..671) (11,53)-(11,73)

System.InvalidCastException: Unable to cast object of type 'Microsoft.CodeAnalysis.CSharp.BoundTupleLiteral' to type 'Microsoft.CodeAnalysis.CSharp.BoundConvertedTupleLiteral'.
   at Microsoft.CodeAnalysis.CSharp.CSharpSemanticModel.GetTypeInfoForNode(BoundNode lowestBoundNode, BoundNode highestBoundNode, BoundNode boundNodeForSyntacticParent)
   at Microsoft.CodeAnalysis.CSharp.MemberSemanticModel.GetTypeInfoWorker(CSharpSyntaxNode node, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CSharpSemanticModel.GetConversion(SyntaxNode expression, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.Utilities.SpeculationAnalyzer.ConversionsAreCompatible(SemanticModel originalModel, ExpressionSyntax originalExpression, SemanticModel newModel, ExpressionSyntax newExpression)
   at Microsoft.CodeAnalysis.Shared.Utilities.AbstractSpeculationAnalyzer`9.ImplicitConversionsAreCompatible(TExpressionSyntax originalExpression, TExpressionSyntax newExpression)
   at Microsoft.CodeAnalysis.Shared.Utilities.AbstractSpeculationAnalyzer`9.ReplacementChangesSemanticsForNode(TSyntaxNode currentOriginalNode, TSyntaxNode currentReplacedNode, TSyntaxNode previousOriginalNode, TSyntaxNode previousReplacedNode)
   at Microsoft.CodeAnalysis.Shared.Utilities.AbstractSpeculationAnalyzer`9.ReplacementChangesSemantics(TSyntaxNode currentOriginalNode, TSyntaxNode currentReplacedNode, TSyntaxNode originalRoot, Boolean skipVerificationForCurrentNode)
   at Microsoft.CodeAnalysis.Shared.Utilities.AbstractSpeculationAnalyzer`9.ReplacementChangesSemantics()
   at Microsoft.CodeAnalysis.CSharp.Extensions.CastExpressionSyntaxExtensions.IsUnnecessaryCast(CastExpressionSyntax cast, SemanticModel semanticModel, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.Diagnostics.RemoveUnnecessaryCast.CSharpRemoveUnnecessaryCastDiagnosticAnalyzer.IsUnnecessaryCast(SemanticModel model, SyntaxNode node, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Diagnostics.RemoveUnnecessaryCast.RemoveUnnecessaryCastDiagnosticAnalyzerBase`1.TryRemoveCastExpression(SemanticModel model, SyntaxNode node, Diagnostic& diagnostic, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Diagnostics.RemoveUnnecessaryCast.RemoveUnnecessaryCastDiagnosticAnalyzerBase`1.<Initialize>b__6_0(SyntaxNodeAnalysisContext nodeContext)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.<>c__DisplayClass42_1`1.<ExecuteSyntaxNodeAction>b__1()
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.ExecuteAndCatchIfThrows_NoLock(DiagnosticAnalyzer analyzer, Action analyze, Nullable`1 info)
-----
'.
```

**Comment by Pilchie** [comment]
Is this something that the IDE is doing wrong in constructing the tree when we rewrite?

**Comment by gafter** [comment]
@Pilchie It could be. If so it would be nice of the compiler to handle it more gracefully.

**Comment by jcouv** [comment]
I feel this is related to https://github.com/dotnet/roslyn/issues/18738 (a crash in GetTypeInfo related to having a BoundTupleLiteral when we expect a BoundConvertedTupleLiteral).
There is a hole in the conversion logic which lets it return a BoundTupleLiteral (in some error case), when it should always return a BoundConvertedTupleLiteral.

**Comment by jcouv** [comment]
Thanks for reporting the issue. 

The semantic model code was fixed to handle this case (BoundTupleLiteral survives, whereas previously we expected it should be upgraded to BoundConvertedTupleLiteral).
Fix is #19679 and will ship in Visual Studio 2017 version 15.3.
